File: framework/languages/java/samples/java/TicTacToe/Server/src/main/java/systems/zlink/samples/tictactoe/server/TicTacToeServerHostFactory.java
package systems.zlink.samples.tictactoe.server;

import org.springframework.context.ConfigurableApplicationContext;
import systems.zlink.samples.tictactoe.server.api.ApiServerHostFactory;
import systems.zlink.samples.tictactoe.server.configuration.SampleSettings;
import systems.zlink.samples.tictactoe.server.play.PlayServerHostFactory;

public final class TicTacToeServerHostFactory {
    private TicTacToeServerHostFactory() {
    }

    public static AutoCloseable start(SampleSettings settings) {
        SampleSettings.setCurrent(settings);
        ConfigurableApplicationContext play = null;
        ConfigurableApplicationContext api = null;
        try {
            play = PlayServerHostFactory.start(settings);
            api = ApiServerHostFactory.start(settings);
            return new ServerHost(play, api);
        } catch (RuntimeException error) {
            close(api);
            close(play);
            throw error;
        }
    }

    private record ServerHost(
        ConfigurableApplicationContext play,
        ConfigurableApplicationContext api) implements AutoCloseable {
        @Override
        public void close() {
            TicTacToeServerHostFactory.close(api);
            TicTacToeServerHostFactory.close(play);
        }
    }

    private static void close(ConfigurableApplicationContext context) {
        if (context != null) {
            context.close();
        }
    }
}

File: framework/languages/java/samples/kotlin/TicTacToe/Server/src/main/kotlin/systems/zlink/samples/kotlin/tictactoe/server/TicTacToeServerHostFactory.kt
package systems.zlink.samples.kotlin.tictactoe.server

import org.springframework.context.ConfigurableApplicationContext
import systems.zlink.samples.kotlin.tictactoe.server.api.ApiServerHostFactory
import systems.zlink.samples.kotlin.tictactoe.server.configuration.SampleSettings
import systems.zlink.samples.kotlin.tictactoe.server.play.PlayServerHostFactory

object TicTacToeServerHostFactory {
    fun start(settings: SampleSettings): AutoCloseable {
        var play: ConfigurableApplicationContext? = null
        var api: ConfigurableApplicationContext? = null
        try {
            play = PlayServerHostFactory.start(settings)
            api = ApiServerHostFactory.start(settings)
            return ServerHost(play, api)
        } catch (error: RuntimeException) {
            api?.close()
            play?.close()
            throw error
        }
    }

    private class ServerHost(
        private val play: ConfigurableApplicationContext?,
        private val api: ConfigurableApplicationContext?,
    ) : AutoCloseable {
        override fun close() {
            api?.close()
            play?.close()
        }
    }
}
